Stdenv setup misc cleanups

Ericson2314 opened the pull request:
###### Motivation for this change

@grahamc showed me shellcheck, which I am now quite fond of. In order to tightening things up so https://github.com/NixOS/nixpkgs/pull/26805 is easier to debug, I did these clean ups.

I left the `# shellcheck disable=...` in there in hopes we can eventually make it a test. A lot of this quoting doesn't by us much since other things block store paths with whitespace, but shellcheck can also make switching to`set -u` feasible, which is definitely worthy of static analysis.

This is easiest to review commit-by-commit:

1. Use associative arrays to accumulate deps. Then instead of doing huge pattern match just check the key, and do get out deps iterator through keys.
2. Shell check
3. Few style + error handling things on `substitue`

###### Things done

Rebuilt Linux and Darwin stdenvs.

Ericson2314 committed:
stdenv-setup: Make the package accumulators associative arrays instead of strings

This is generally cleaner: less eval, less worrying about separators,
and probably also faster. I got the idea from that python wrapper
script.

Ericson2314 committed:
stdenv-setup: Misc improvements as directed by ShellCheck

I took some liberties with the flags-echoing code to make it more
concise and correct. Also, a few warnings in findInputs and friends I
skipped because I am going to rewrite those anyways.

Thanks @grahamc for telling me about this great linter!

Ericson2314 committed:
stdenv-setup: Clean up 'substitute()' for style and error handling

It now blows up on null byte in file (rather than silently truncating),
and invalid arguments (rather than silently skipping).

copumpkin commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -211,7 +211,7 @@ stdenv.mkDerivation ({
 
     # nativePkgs defined in stdenv/setup.hs
-    for p in $nativePkgs; do
+    for p in "${!nativePkgs[@]}"; do
```
Does this work? Isn't the ${} escaping into nix?

Ericson2314 replied:
Weird you would think, but I did build it.

joachifm commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -422,50 +421,52 @@ fi
-    local input="$1"
-    local output="$2"
+    local input=$1; shift
+    local output=$1; shift
```
`shift` also takes a numeric argument, so you can do e.g., `local foo=$1 bar=$2; shift 2` to achieve the same effect.

Ericson2314 replied:
I short of like the "one variable, one shift" correspondence, but I'd be happy to change it.

joachifm replied:
Looks a bit odd to me, but it's entirely superficial :)

Ericson2314 replied:
Ah I can avoid binding `p` doing this. Swayed :).

edolstra commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -422,50 +421,55 @@ fi
-    local input="$1"
-    local output="$2"
+    local input=$1
+    local output=$2
```
Why no quotes here?

Ericson2314 replied:
Bash parsed the same either way! I learned this and was quite surprised.

Ericson2314 replied:
I could leave as is, but I figured given it didn't matter I should pick one and be consistent. I could normalize the other way if you like.

Ericson2314 replied:
```bash
$ set -x; x='asdf asdf'; y=$x; z="$x"
+ set -x
+ x='asdf asdf'
+ y='asdf asdf'
+ z='asdf asdf'
```

joachifm replied:
This is covered under PARAMETERS in the bash manual. Word splitting & pathname expansion is not performed on assignment.

dezgeg replied:
I don't find this an improvement, it's just forces people to learn another special case in bash syntax instead of using the obviously safe variant.

copumpkin replied:
Agree with @dezgeg here. Shell quoting is such a pain and source of subtle bugs already; although it may be technically correct, why not be explicit when we can do so trivially? This particular gun might have a fancy high tech anti-foot-shot protection device in it, but I'd rather just point it away from my feet.

vcunat replied:
Uh, I hate how the bash language is so complex.

Ericson2314 replied:
I basically got bash Stockholm syndrome writing this stuff. Happy to quote everything for consistency.

edolstra commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -727,17 +742,23 @@ configurePhase() {
     runHook preBuild
 
-    if [ -z "$makeFlags" ] && ! [ -n "$makefile" -o -e "Makefile" -o -e "makefile" -o -e "GNUmakefile" ]; then
+    if [ -z "$makeFlags" ] && ! [[ -n "$makefile" || -e "Makefile" || -e "makefile" || -e "GNUmakefile" ]]; then
```
Can't this be written as a single `[[...]]` test? Something like
```
if [[ -z $makeFlags && ! ( -n $makefile || ... ) ]]; then
```
Note that quotes are not necessary inside `[[...]]`.

Ericson2314 replied:
Yes it can. Will do.

edolstra commented:
Looks good to me!

edolstra commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -275,17 +277,14 @@ runHook addInputsHook
-    eval $var="'${!var} $pkg '"
+    # Stop if we've already added this one
+    [[ -z "${varDeref["$pkg"]}" ]] || return 0
+    varDeref["$pkg"]=1
```
This is great because it fixes the O(n^2) performance setting up the environment!

Ericson2314 committed:
stdenv-setup: Combine [[ .. ]] && [[ .. ]] into one [[ .. && .. ]]

Also remove useless quotes on same line

Ericson2314 committed:
stdenv-setup: Remove useless quotes

foo=$1 surprisingly doesn't need quotes in Bash. Word splits are only
syntactic in string variable (not array var!) assignments.

Ericson2314 commented:
Doing another build with those changes, and then will merge.

dezgeg commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -765,10 +792,17 @@ installPhase() {
-        $installFlags "${installFlagsArray[@]}"
+        $installFlags "${installFlagsArray[@]}")
+
+    printf 'install flags:'
```
Why is this multiple commands?

Ericson2314 replied:
@dezgeg it hash to be:

1. print the name of the flags
2. repeat the format string (*un*quoted space + quoted arg) for each argument
3. close the line

printf repeats the format string until all the args are gone, so it is essential the repeated format string contain one a single substitution.

Ericson2314 replied:
I could add a helper function and comments for this :).

dezgeg replied:
Ah.

dezgeg commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -422,50 +421,55 @@ fi
+    local content
+    # read returns non-0 on EOF, so we want read to fail
+    if IFS='' read -r -N 0 content < "$input"; then
+        echo "${FUNCNAME[0]}(): ERROR: File \"$input\" has null bytes, won't process" >&2
```
`substitute` is documented to work on binaries in the manual:
> It supports performing substitutions on binary files (such as executables), though there you’ll probably want to make sure that the replacement string is as long as the replaced string.

Ericson2314 replied:
@dezgeg Woah. Well the manual is definitely wrong about that. We can use `mapfile` and carefully `printf` the missing null bytes, but I think that should be a followup PR.

Ericson2314 committed:
stdenv-setup: Pull out and explain 3-part printing of commands

@Dezgeg made the good point that the reasons for doing this were not at
all intuitive.
